File: src/main/java/com/multicampus/gamesungcoding/a11ymarketserver/feature/auth/service/AuthService.java
package com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.service;

import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataDuplicatedException;
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataNotFoundException;
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.InvalidRequestException;
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.UserNotFoundException;
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.dto.JwtResponse;
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.provider.JwtTokenProvider;
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.service.RefreshTokenService;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.dto.*;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.status.CheckExistsStatus;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.dto.UserResponse;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.UserRole;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserOauthLinksRepository;
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Collections;
import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final RefreshTokenService refreshTokenService;
    private final UserDetailsService userDetailsService;
    private final JwtTokenProvider jwtTokenProvider;
    private final AuthenticationManager authenticationManager;
    private final UserOauthLinksRepository userOauthLinksRepository;

    public LoginResponse login(LoginRequest dto) {

        Users user = userRepository.findByUserEmail(dto.email())
                .orElseThrow(() -> new UserNotFoundException("이메일 또는 비밀번호가 올바르지 않습니다."));

        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(dto.email(), dto.password())
        );

        return LoginResponse.fromEntityAndTokens(
                user,
                jwtTokenProvider.createAccessToken(authentication),
                refreshTokenService.createRefreshToken(authentication)
        );
    }

    public LoginResponse loginRefresh(String refreshToken) {

        // get refresh token from DB and verifying validation
        var user = getUserByRefreshToken(refreshToken);
        var userDetails = userDetailsService.loadUserByUsername(user.getUserEmail());

        var newAuthentication = new UsernamePasswordAuthenticationToken(
                userDetails,
                null,
                userDetails.getAuthorities()
        );

        String newAccessToken = jwtTokenProvider.createAccessToken(newAuthentication);

        return LoginResponse.fromEntityAndTokens(
                user,
                newAccessToken,
                refreshToken
        );
    }

    public LoginResponse getUserInfo(UUID userId) {
        Users user = userRepository.findById(userId)
                .orElseThrow(() -> new UserNotFoundException("유효하지 않은 사용자입니다."));

        List<GrantedAuthority> authorities = Collections.singletonList(
                new SimpleGrantedAuthority(user.getUserRole().name())
        );

        Authentication authentication = new UsernamePasswordAuthenticationToken(
                user.getUserEmail(),
                null,
                authorities
        );

        return LoginResponse.fromEntityAndTokens(
                user,
                jwtTokenProvider.createAccessToken(authentication),
                refreshTokenService.createRefreshToken(authentication)
        );
    }

    @Transactional
    public JwtResponse reissueToken(String refreshToken) {

        var user = getUserByRefreshToken(refreshToken);
        var userDetails = userDetailsService.loadUserByUsername(user.getUserEmail());

        var newAuthentication = new UsernamePasswordAuthenticationToken(
                userDetails,
                null,
                userDetails.getAuthorities()
        );

        String newAccessToken = jwtTokenProvider.createAccessToken(newAuthentication);
        return new JwtResponse(newAccessToken, refreshToken);
    }

    @Transactional
    public UserResponse join(JoinRequest dto) {

        // 이메일 중복 체크
        if (userRepository.existsByUserEmail(dto.userEmail())) {
            throw new DataDuplicatedException("이미 존재하는 이메일입니다.");
        }

        // 비밀번호 암호화
        String encodedPwd = passwordEncoder.encode(dto.userPass());

        Users user = Users.builder()
                .userEmail(dto.userEmail())
                .userPass(encodedPwd)
                .userName(dto.userName())
                .userNickname(dto.userNickname())
                .userPhone(dto.userPhone())
                .userRole(UserRole.USER)
                .build();
        return UserResponse.fromEntity(userRepository.save(user));
    }

    @Transactional
    public UserResponse kakaoJoin(UUID userOauthLinkId, KakaoSignUpRequest dto) {
        if (userRepository.existsByUserEmail(dto.userEmail())) {
            throw new DataDuplicatedException("이미 존재하는 이메일입니다.");
        }

        var oauthLink = userOauthLinksRepository.findById(userOauthLinkId)
                .orElseThrow(() -> new DataNotFoundException("OAuth link not found for ID: " + userOauthLinkId));
        if (oauthLink.getUser() != null) {
            throw new InvalidRequestException("이미 가입된 OAuth 링크입니다.");
        }

        Users user = userRepository.save(
                Users.builder()
                        .userEmail(dto.userEmail())
                        .userName(dto.userName())
                        .userNickname(dto.userNickname())
                        .userRole(UserRole.USER)
                        .build());
        oauthLink.updateUser(user);
        oauthLink = userOauthLinksRepository.save(oauthLink);

        return UserResponse.fromEntity(oauthLink.getUser());
    }

    // 이메일 중복 체크 API용
    public CheckExistsResponse isEmailDuplicate(String email) {
        if (userRepository.existsByUserEmail(email)) {
            return new CheckExistsResponse(CheckExistsStatus.UNAVAILABLE);
        } else {
            return new CheckExistsResponse(CheckExistsStatus.AVAILABLE);
        }
    }

    public CheckExistsResponse isPhoneDuplicate(String phone) {
        if (userRepository.existsByUserPhone(phone)) {
            return new CheckExistsResponse(CheckExistsStatus.UNAVAILABLE);
        } else {
            return new CheckExistsResponse(CheckExistsStatus.AVAILABLE);
        }
    }

    public CheckExistsResponse isNicknameDuplicate(String nickname) {
        if (userRepository.existsByUserNickname(nickname)) {
            return new CheckExistsResponse(CheckExistsStatus.UNAVAILABLE);
        } else {
            return new CheckExistsResponse(CheckExistsStatus.AVAILABLE);
        }
    }

    public void logout(String userEmail) {
        userRepository.findByUserEmail(userEmail)
                .orElseThrow(() -> new InvalidRequestException("유효하지 않은 사용자입니다."));
        refreshTokenService.deleteRefreshToken(userEmail);
    }

    private Users getUserByRefreshToken(String refreshToken) {
        var dbToken = refreshTokenService.verifyRefreshToken(refreshToken);
        return userRepository.findById(dbToken.getUser().getUserId())
                .orElseThrow(() ->
                        new DataNotFoundException("User not found for: " + dbToken.getUser().getUserEmail()));
    }
}

File: src/main/kotlin/com/multicampus/gamesungcoding/a11ymarketserver/feature/auth/service/AuthService.kt
package com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.service

import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataDuplicatedException
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.DataNotFoundException
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.InvalidRequestException
import com.multicampus.gamesungcoding.a11ymarketserver.common.exception.UserNotFoundException
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.dto.JwtResponse
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.provider.JwtTokenProvider
import com.multicampus.gamesungcoding.a11ymarketserver.common.jwt.service.RefreshTokenService
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.dto.*
import com.multicampus.gamesungcoding.a11ymarketserver.feature.auth.status.CheckExistsStatus
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.dto.UserResponse
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.UserRole
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.entity.Users
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserOauthLinksRepository
import com.multicampus.gamesungcoding.a11ymarketserver.feature.user.repository.UserRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val refreshTokenService: RefreshTokenService,
    private val userDetailsService: UserDetailsService,
    private val jwtTokenProvider: JwtTokenProvider,
    private val authenticationManager: AuthenticationManager,
    private val userOauthLinksRepository: UserOauthLinksRepository,
) {
    @Transactional(readOnly = true)
    fun login(dto: LoginRequest): LoginResponse {
        val user = userRepository.findByUserEmail(dto.email)
            ?: throw UserNotFoundException("이메일 또는 비밀번호가 올바르지 않습니다.")

        val authentication = authenticationManager.authenticate(
            UsernamePasswordAuthenticationToken(dto.email, dto.password)
        )

        return LoginResponse.fromEntityAndTokens(
            user,
            jwtTokenProvider.createAccessToken(authentication),
            refreshTokenService.createRefreshToken(authentication)
        )
    }

    @Transactional(readOnly = true)
    fun loginRefresh(refreshToken: String): LoginResponse {
        // get refresh token from DB and verifying validation
        val user = getUserByRefreshToken(refreshToken)
        val userDetails = userDetailsService.loadUserByUsername(user.userEmail)

        val newAuthentication = UsernamePasswordAuthenticationToken(
            userDetails,
            null,
            userDetails.authorities
        )

        return LoginResponse.fromEntityAndTokens(
            user,
            jwtTokenProvider.createAccessToken(newAuthentication),
            refreshToken
        )
    }

    @Transactional(readOnly = true)
    fun getUserInfo(userId: UUID): LoginResponse {
        val user = userRepository.findByIdOrNull(userId)
            ?: throw UserNotFoundException("유효하지 않은 사용자입니다.")

        val authentication: Authentication = UsernamePasswordAuthenticationToken(
            user.userEmail,
            null,
            listOf(SimpleGrantedAuthority(user.userRole.name))
        )

        return LoginResponse.fromEntityAndTokens(
            user,
            jwtTokenProvider.createAccessToken(authentication),
            refreshTokenService.createRefreshToken(authentication)
        )
    }

    @Transactional
    fun reissueToken(refreshToken: String): JwtResponse {
        val user = getUserByRefreshToken(refreshToken)
        val userDetails = userDetailsService.loadUserByUsername(user.userEmail)

        val newAuthentication = UsernamePasswordAuthenticationToken(
            userDetails,
            null,
            userDetails.authorities
        )

        return JwtResponse(
            jwtTokenProvider.createAccessToken(newAuthentication),
            refreshToken
        )
    }

    @Transactional
    fun join(dto: JoinRequest): UserResponse {
        // 이메일 중복 체크
        if (userRepository.existsByUserEmail(dto.userEmail)) {
            throw DataDuplicatedException("이미 존재하는 이메일입니다.")
        }

        val user = Users.builder()
            .userEmail(dto.userEmail)
            // 비밀번호 암호화
            .userPass(passwordEncoder.encode(dto.userPass))
            .userName(dto.userName)
            .userNickname(dto.userNickname)
            .userPhone(dto.userPhone)
            .userRole(UserRole.USER)
            .build()

        return UserResponse.fromEntity(userRepository.save(user))
    }

    @Transactional
    fun kakaoJoin(userOauthLinkId: UUID, dto: KakaoSignUpRequest): UserResponse {
        if (userRepository.existsByUserEmail(dto.userEmail)) {
            throw DataDuplicatedException("이미 존재하는 이메일입니다.")
        }

        var oauthLink = userOauthLinksRepository.findByIdOrNull(userOauthLinkId)
            ?: throw DataNotFoundException("OAuth link not found for ID: $userOauthLinkId")
        if (oauthLink.user != null) {
            throw InvalidRequestException("이미 가입된 OAuth 링크입니다.")
        }

        val user = userRepository.save(
            Users.builder()
                .userEmail(dto.userEmail)
                .userName(dto.userName)
                .userNickname(dto.userNickname)
                .userRole(UserRole.USER)
                .build()
        )

        oauthLink.updateUser(user)
        oauthLink = userOauthLinksRepository.save(oauthLink)

        return UserResponse.fromEntity(oauthLink.user)
    }

    // 이메일 중복 체크 API용
    @Transactional(readOnly = true)
    fun isEmailDuplicate(email: String): CheckExistsResponse =
        CheckExistsResponse(
            if (userRepository.existsByUserEmail(email)) CheckExistsStatus.UNAVAILABLE
            else CheckExistsStatus.AVAILABLE
        )

    @Transactional(readOnly = true)
    fun isPhoneDuplicate(phone: String): CheckExistsResponse =
        CheckExistsResponse(
            if (userRepository.existsByUserPhone(phone)) CheckExistsStatus.UNAVAILABLE
            else CheckExistsStatus.AVAILABLE
        )


    @Transactional(readOnly = true)
    fun isNicknameDuplicate(nickname: String): CheckExistsResponse =
        CheckExistsResponse(
            if (userRepository.existsByUserNickname(nickname)) CheckExistsStatus.UNAVAILABLE
            else CheckExistsStatus.AVAILABLE
        )

    @Transactional
    fun logout(userEmail: String) {
        userRepository.findByUserEmail(userEmail)
            ?: throw InvalidRequestException("유효하지 않은 사용자입니다.")
        refreshTokenService.deleteRefreshToken(userEmail)
    }

    private fun getUserByRefreshToken(refreshToken: String): Users {
        val dbToken = refreshTokenService.verifyRefreshToken(refreshToken)
        return userRepository.findByIdOrNull(dbToken.user.userId)
            ?: throw DataNotFoundException("User not found for: " + dbToken.user.userEmail)
    }
}